---
title: "Bacillus proteogenomics report"
author: "`r Sys.info()[['user']]`"
date: "`r format(x = Sys.time(), '%d %B %Y %T')`"
output:
    ioslides_presentation:
        self_contained: FALSE
        css: "C:/Users/kxmna01/Documents/GitHub/Miscellaneous/R/Presentation/ioslides_custom.css"
params:
    evidences: NA
    pept.posit: NA
    levenshtein: NA
    pheno: NA
    bsu.ideo: NA
    ref.grange: NA
    novel.grange: NA
    pep_loc: NA
    manual.annot: NA
---



## Overview: feature counts per type

``` {r overview, eval = TRUE, echo = FALSE, warning = FALSE}

# Compute count of evidence per group type
data <- table(params[["evidences"]]$group) %>%
    data.frame(., type = "evidence", stringsAsFactors = FALSE) %>%
    set_names(c("group", "Freq", "type"))

# Compute count of unique peptide per group type
data <- params[["evidences"]] %>%
    dplyr::select(., Sequence, group) %>%
    dplyr::group_by(., group) %>%
    dplyr::summarise(., Freq = n_distinct(Sequence)) %>%
    data.frame(., type = "peptide", stringsAsFactors = FALSE) %>%
    rbind(data, .)

# Histogram of feature count per group type
pl <- plots_hist(
    data = data,
    key = "type",
    value = "Freq",
    group = "group",
    fill = "group",
    main = "Number of evidence/peptide",
    transf = "log10",
    label = "Freq")
pl[[1]]

```



## Overview: PEP per type

``` {r overview2, eval = TRUE, echo = FALSE, warning = FALSE}

# Plot evidence PEP distribution
pl <- plots_violon(
    data = params[["evidences"]],
    key = "group",
    value = "PEP",
    fill = "lightblue",
    colour = "black",
    main = "PEP comparison",
    zoom = c(-0.001, quantile(params[["evidences"]]$PEP, 0.95)[[1]]))
pl[[1]]

```



## Overview: PEP per type (zoom)

``` {r overview2bis, eval = TRUE, echo = FALSE, warning = FALSE}

# Plot evidence PEP distribution
pl <- plots_violon(
    data = params[["evidences"]],
    key = "group",
    value = "PEP",
    fill = "lightblue",
    colour = "black",
    main = "PEP comparison",
    zoom = c(-0.001, quantile(params[["evidences"]]$PEP, 0.95)[[1]]))
pl[[2]]

```



## Overview: Score per type

``` {r overview3, eval = TRUE, echo = FALSE, warning = FALSE}

# Plot evidence score distribution
pl <- plots_violon(
    data = params[["evidences"]],
    key = "group",
    value = "Score",
    fill = "lightblue",
    colour = "black",
    main = "Score comparison")
pl[[1]]

```



## Overview: Mass error per type

``` {r overview4, eval = TRUE, echo = FALSE, warning = FALSE}

# Calculate quantile values of evidence mass error
data <- params[["evidences"]] %>%
    dplyr::select(., group, `Mass Error [ppm]`) %>%
    base::as.data.frame(., stringsAsFactors = FALSE) %>%
    set_colnames(c("group", "mass_error"))

# Plot evidence mass error distribution
pl <- plots_violon(
    data = data,
    key = "group",
    value = "mass_error",
    fill = "lightblue",
    colour = "black",
    main = "Mass error (ppm) comparison")
pl[[1]]

```



## Novel features count

``` {r NovelCount, eval = TRUE, echo = FALSE, warning = FALSE}

# Compute novel feature count (peptide and ORF) in total
data <- params[["pept.posit"]] %>%
    dplyr::summarise(
        .,
        type = "All novel",
        Number_peptide = n_distinct(Sequence),
        Number_ORF = n_distinct(ORF)) %>%
    tidyr::gather(
        data = ., key = "feature", value = "Count", -type) %>%
    base::as.data.frame(., stringsAsFactors = FALSE)

# Compute novel feature count (peptide and ORF) that pass PEP filter 
data <- params[["pept.posit"]] %>%
    dplyr::filter(., ReasonNovel != "PEPfilter") %>%
    dplyr::summarise(
        .,
        type = "PEP filtered",
        Number_peptide = n_distinct(Sequence),
        Number_ORF = n_distinct(ORF)) %>%
    tidyr::gather(
        data = ., key = "feature", value = "Count", -type) %>%
    base::as.data.frame(., stringsAsFactors = FALSE) %>%
    rbind(data, .)

# Plot the novel feature count
pl <- plots_hist(
    data = data,
    key = "feature",
    value = "Count",
    group = "type",
    fill = "type",
    main = "Novel peptide/ORF count",
    label = "Count")
pl[[1]]

```



## Levenshtein distance for SAV peptide

``` {r levenshtein, eval = TRUE, echo = FALSE, warning = FALSE}

# Compute the peptide count per levenshtein distance
data <- params[["levenshtein"]] %>%
    dplyr::select(., Sequence, leven) %>%
    unique(.) %>%
    dplyr::group_by(., leven) %>%
    dplyr::summarise(., count = n()) %>%
    base::as.data.frame(., stringsAsFactors = FALSE)

# Plot the levenshtein distance count
pl <- plots_hist(
    data = data,
    key = "leven",
    value = "count",
    group = "leven",
    fill = "leven",
    main = "Levenshtein distance density",
    textsize = 15,
    label = "count")
pl[[1]]

```



## Novel peptide reason explanation

``` {r NovelReason, eval = TRUE, echo = FALSE, warning = FALSE}

# Compute the peptide count per novelty reasons
data <- params[["pept.posit"]] %>%
    dplyr::filter(., ReasonNovel != "PEPfilter") %>%
    dplyr::group_by(., ReasonNovel) %>%
    dplyr::summarise(
        .,
        Number_peptide = n_distinct(Sequence),
        Number_ORF = n_distinct(ORF)) %>%
    tidyr::gather(
        data = ., key = "feature", value = "Count", -ReasonNovel) %>%
    base::as.data.frame(., stringsAsFactors = FALSE)

# Plot the feature count per novelty reasons
pl <- plots_hist(
    data = data,
    key = "feature",
    value = "Count",
    group = "ReasonNovel",
    fill = "ReasonNovel",
    main = "Peptide novelty reasons",
    label = "Count")
pl[[1]]

```



## ORF density per novel peptide count

``` {r PepCount, eval = TRUE, echo = FALSE, warning = FALSE}

# Compute count of unique peptide per ORF (to see which ORF
# have multiple peptide evidence)
novelty <- unique(params[["pept.posit"]]$ReasonNovel)[-1]
data <- data.frame()
for (key in novelty) {
    
    filt <- params[["pept.posit"]] %>%
        dplyr::filter(., ReasonNovel == key) %>%
        .[["ORF"]]
    data <- params[["pept.posit"]] %>%
        dplyr::filter(., ORF %in% filt) %>%
        dplyr::group_by(., ORF) %>%
        dplyr::summarise(., Unique_peptide_count = n_distinct(Sequence)) %>%
        dplyr::group_by(., Unique_peptide_count) %>%
        dplyr::summarise(., ORF_count = n_distinct(ORF)) %>%
        base::data.frame(., Novelty = key, stringsAsFActors = FALSE) %>%
        rbind(data, .)
    
}

# Plot the ORF frequency per matching peptide count
pl <- plots_hist(
    data = data,
    key = "Unique_peptide_count",
    value = "ORF_count",
    group = "Novelty",
    fill = "Novelty",
    main = "Novel ORF density per novel peptide count",
    label = "ORF_count")
pl[[1]]

```



## Peptide proportion per condition

``` {r PepCondi, eval = TRUE, echo = FALSE, warning = FALSE}

# Calculate number of unique peptide per group and conditions
pep.cond <- params[["evidences"]] %>%
    dplyr::left_join(
        x = .,
        y = params[["pheno"]] %>% dplyr::select(., Name, Conditions, Modification),
        by = c("Raw file" = "Name")) %>%
    dplyr:::group_by(., group, Conditions) %>%
    dplyr::summarise(., count = n_distinct(Sequence)) %>%
    dplyr::ungroup(.) %>%
    dplyr::group_by(., group) %>%
    dplyr::mutate(., perc = count / sum(count)) %>%
    dplyr::select(., -count) %>%
    base::as.data.frame(., stringsAsFactors = FALSE)

# Compute the condition repartition for the overall experiment
pep.cond <- params[["pheno"]] %>%
    dplyr:::group_by(., Conditions) %>%
    dplyr::summarise(., count = n_distinct(Name)) %>%
    dplyr::ungroup(.) %>%
    dplyr::mutate(
        .,
        group = "Condition per samples",
        perc = count / sum(count)) %>%
    dplyr::select(., -count) %>%
    dplyr::group_by(., group) %>%
    base::as.data.frame(., stringsAsFactors = FALSE) %>%
    base::rbind(pep.cond, .)

# Line plot of the peptide proportion per group and condition
pl <- plots_line(
    data = pep.cond,
    key = "Conditions",
    value = "perc",
    group = "group",
    fill = "group",
    colour = "group",
    main = "Peptide proportion per group and condition",
    textsize = 15)
pl[[1]]

```



## Genomic mapping of ORF

``` {r Circos, eval = TRUE, echo = FALSE, warning = FALSE}

# Use ggbio extension to plot ORF location on genome as a circos graph
colou <- c("#4682B4", "#BD5E5E", "#437A3C", "#F57C36", "#D58DEB", "#B2B83F")
pl1 <- ggplot() +
    ggtitle(label = "Reference ORF") +
    layout_circle(
        params[["bsu.ideo"]], geom = "ideo", fill = "gray70",
        radius = 30, trackWidth = 3) +
    layout_circle(
        params[["bsu.ideo"]], geom = "scale", size = 2, radius = 33, trackWidth = 2) +
    layout_circle(
        params[["bsu.ideo"]], geom = "text", aes(label = seqnames),
        angle = 0, radius = 36, trackWidth = 5) +
    layout_circle(
        subset(x = params[["ref.grange"]], frame == 1), geom = "rect", color = colou[1],
        radius = 26, trackWidth = 3) +
    layout_circle(
        subset(x = params[["ref.grange"]], frame == 2), geom = "rect", color = colou[2],
        radius = 23, trackWidth = 3) +
    layout_circle(
        subset(x = params[["ref.grange"]], frame == 3), geom = "rect", color = colou[3],
        radius = 20, trackWidth = 3) +
    layout_circle(
        subset(x = params[["ref.grange"]], frame == -1), geom = "rect", color = colou[4],
        radius = 17, trackWidth = 3) +
    layout_circle(
        subset(x = params[["ref.grange"]], frame == -2), geom = "rect", color = colou[5],
        radius = 14, trackWidth = 3) +
    layout_circle(
        subset(x = params[["ref.grange"]], frame == -3), geom = "rect", color = colou[6],
        radius = 11, trackWidth = 3)

# Use ggbio extension to plot ORF location on genome as a circos graph
pl2 <- ggplot() +
    ggtitle(label = "Novel ORF") +
    layout_circle(
        params[["bsu.ideo"]], geom = "ideo", fill = "gray70",
        radius = 30, trackWidth = 3) +
    layout_circle(
        params[["bsu.ideo"]], geom = "scale", size = 2, radius = 33, trackWidth = 2) +
    layout_circle(
        params[["bsu.ideo"]], geom = "text", aes(label = seqnames),
        angle = 0, radius = 36, trackWidth = 5) +
    layout_circle(
        subset(x = params[["novel.grange"]], frame == 1), geom = "rect", color = colou[1],
        radius = 26, trackWidth = 3) +
    layout_circle(
        subset(x = params[["novel.grange"]], frame == 2), geom = "rect", color = colou[2],
        radius = 23, trackWidth = 3) +
    layout_circle(
        subset(x = params[["novel.grange"]], frame == 3), geom = "rect", color = colou[3],
        radius = 20, trackWidth = 3) +
    layout_circle(
        subset(x = params[["novel.grange"]], frame == -1), geom = "rect", color = colou[4],
        radius = 17, trackWidth = 3) +
    layout_circle(
        subset(x = params[["novel.grange"]], frame == -2), geom = "rect", color = colou[5],
        radius = 14, trackWidth = 3) +
    layout_circle(
        subset(x = params[["novel.grange"]], frame == -3), geom = "rect", color = colou[6],
        radius = 11, trackWidth = 3)
pl <- list("Known" = pl1, "Novel" = pl2)

# Arrange the obtained plots on page
ml <- marrangeGrob(pl, nrow = 1, ncol = 2, top = NULL)
ml

```



## Genomic coverage

``` {r GenomCover, eval = TRUE, echo = FALSE, warning = FALSE, fig.width = 5, fig.height = 5}

# Get all chromosome nucleotide position
chrom.nuc <- params[["bsu.ideo"]]@ranges[[1]]

# Get all protein-coding associated nucleotide position
coding.nuc <- lapply(X = params[["ref.grange"]]@ranges, FUN = function(x) {
    x
})
names(coding.nuc) <- params[["ref.grange"]]@elementMetadata@listData$UniProtKBID

# Get all expressed protein associated nucleotide position
exprs.id <- strsplit(x = params[["evidences"]]$Proteins, split = ";") %>%
    unlist(.) %>%
    unique(.)
exprs.nuc <- coding.nuc[names(coding.nuc) %in% exprs.id]

# Get all peptide associated nucleotide position
tmp <- params[["pep_loc"]][
    params[["pep_loc"]]$Proteins %in% names(coding.nuc) &
        !is.na(params[["pep_loc"]]$start), ]
cover.nuc <- tmp %>%
    apply(
    X = .,
    MARGIN = 1,
    FUN = function(x) {
        
        val <- c(x["start"]:x["end"])
        tmp <- coding.nuc[[x["Proteins"]]][val]
        tmp
        
    })
names(cover.nuc) <- paste(tmp$Sequence, tmp$Proteins, sep = "_")

# Format data into a standard plotting dataframe
toplot <- data.frame(
    Param = paste(
        "Chromosome", round(length(chrom.nuc) / 1000000, 1),
        "Mb", sep = " "),
    Position = as.integer(unique(chrom.nuc)),
    stringsAsFactors = FALSE)
toplot <- data.frame(
    Param = paste(
        "Protein-coding",
        round(length(unique(unlist(coding.nuc))) / 1000000, 1),
        "Mb", sep = " "),
    Position = as.integer(unique(unlist(coding.nuc))),
    stringsAsFactors = FALSE) %>%
    base::rbind(toplot, .)
toplot <- data.frame(
    Param = paste(
        "Expressed protein",
        round(length(unique(unlist(exprs.nuc))) / 1000000, 1),
        "Mb", sep = " "),
    Position = as.integer(unique(unlist(exprs.nuc))),
    stringsAsFactors = FALSE) %>%
    base::rbind(toplot, .)
toplot <- data.frame(
    Param = paste(
        "Detected peptide",
        round(length(unique(unlist(cover.nuc))) / 1000000, 1),
        "Mb", sep = " "),
    Position = as.integer(unique(unlist(cover.nuc))),
    stringsAsFactors = FALSE) %>%
    base::rbind(toplot, .)

# Plot a square venn diagram of chromosome coverage
plot.new()
rect(
    xleft = 0,
    ybottom = 0,
    xright = 1,
    ytop = 1,
    border = "red",
    lwd = 2)
text(
    x = 0.015,
    y = 1.03,
    labels = paste(
        "Chromosome", round(length(chrom.nuc) / 1000000, 1),
        "Mb", sep = " "),
    col = "red", cex = 1.0, adj = 0)
rect(
    xleft = 0.01,
    ybottom = 0.01,
    xright = sqrt(
        length(unique(unlist(coding.nuc))) / length(chrom.nuc)) + 0.01,
    ytop = sqrt(
        length(unique(unlist(coding.nuc))) / length(chrom.nuc)) + 0.01,
    border = "green",
    lwd = 2)
text(
    x = 0.025,
    y = sqrt(
        length(unique(unlist(coding.nuc))) / length(chrom.nuc)) + 0.01 - 0.035,
    labels = paste(
        "Protein-coding",
        round(length(unique(unlist(coding.nuc))) / 1000000, 1),
        "Mb", sep = " "),
    col = "green", cex = 1.0, adj = 0)
rect(
    xleft = 0.02,
    ybottom = 0.02,
    xright = sqrt(
        length(unique(unlist(exprs.nuc))) / length(chrom.nuc)) + 0.02,
    ytop = sqrt(
        length(unique(unlist(exprs.nuc))) / length(chrom.nuc)) + 0.02,
    border = "blue",
    lwd = 2)
text(
    x = 0.035,
    y = sqrt(
        length(unique(unlist(exprs.nuc))) / length(chrom.nuc)) + 0.02 - 0.035,
    labels = paste(
        "Expressed protein",
        round(length(unique(unlist(exprs.nuc))) / 1000000, 1),
        "Mb", sep = " "),
    col = "blue", cex = 1.0, adj = 0)
rect(
    xleft = 0.03,
    ybottom = 0.03,
    xright = sqrt(
        length(unique(unlist(cover.nuc))) / length(chrom.nuc)) + 0.03,
    ytop = sqrt(
        length(unique(unlist(cover.nuc))) / length(chrom.nuc)) + 0.03,
    border = "gold",
    lwd = 2)
text(
    x = 0.045,
    y = sqrt(
        length(unique(unlist(cover.nuc))) / length(chrom.nuc)) + 0.03 + 0.03,
    labels = paste(
        "Detected peptide",
        round(length(unique(unlist(cover.nuc))) / 1000000, 1),
        "Mb", sep = " "),
    col = "gold", cex = 1.0, adj = 0)

```



## Novel peptide manual annotation

``` {r ManualAnnot, eval = TRUE, echo = FALSE, warning = FALSE, fig.width = 5, fig.height = 5}

# Display the table of manual ORF annotation
DT::datatable(
    data = params[["manual.annot"]] %>%
        dplyr::select(
            ., ORF, ManualAnnotation, PeptideCount),
    rownames = FALSE, fillContainer = TRUE)

```


